### DISCUSSION TITLE
Injection operators

### DISCUSSION THREAD

**Discussion by pauleveritt** [Finistere/antidote#55] [Ideas]
*Note: This relates to something mentioned in the props discussion about `current_user` as an injectable function.*

My Hopscotch injector has a useful feature called [operators](https://hopscotch.readthedocs.io/en/latest/injection.html#operators). It lets you write little re-usable callables as part of injection. For example: 

```python
@dataclass()
class GreeterFirstName:
    customer_name: str = get(Customer, attr="first_name")
```

What problems is it trying to solve?

- **Reduce surface area of contract**. Don't ask for the entire `Customer` if you're only using the `title`.
- **Props**. If parent component wants to customize title, they don't have to pass in an entire `Customer`.
- **Persistence and change**. If a customer's description changed, but not its title, then this component instance doesn't need updating.
- **Semantic**. They are so easy to write, you can make domain-specific "verbs" for your framework users.
- **Pipelines**. If you can chain them, you get kind of an RxJS-style pipeline.
- **Testability**. It's a lot easier to test a component when you just have to pass in a simple value.

## Hopscotch Implementation

I had to rewrite this a few times as I made it too magical, didn't handle control flow well, typing was a mess, etc. In essence, the injector has access to the registry and passed it into the operator instance. It's free then to do whatever it wants.

[Here is the implementation](https://github.com/pauleveritt/hopscotch/blob/main/src/hopscotch/operators.py)

This is easier obviously in Antidote as anybody can grab the `world` instance.

**Comment by Finistere** [comment]
Is the following what you need?
```python
from dataclasses import dataclass
from typing import Any

from antidote import inject, injectable, lazy, world


@lazy
def get(dependency: type, *, attr: str) -> Any:
    return getattr(world.get[Any](dependency), attr)


@injectable
class Customer:
    first_name: str = "John"


@injectable
@dataclass
class GreeterFirstName:
    customer_name: str = get(Customer, attr="first_name")
```

**Comment by Finistere** [comment]
> Persistence and change. If a customer's description changed, but not its title, then this component instance doesn't need updating.

In theory, this should be possible with AEP2 (`@state` & `deterministic`)

As a side note, it may have been a mistake to use singleton by default for `@lazy`... :/

**Reply by pauleveritt** [reply]
It will be interesting to see the interaction with @state. It will also be interesting to hear you give me better ideas about state. 😁

But my thing doesn't do any of that, so what you provided is already better. Unless of course `mypy` gets mad. I will give it a try.

**Comment by Finistere** [comment]
By the way, this is better in terms of typing:
```python
from dataclasses import dataclass
from typing import Callable, Type, TypeVar

from antidote import injectable, lazy, world

T = TypeVar('T')
R = TypeVar('R')


@lazy
def get(dependency: Type[T], callback: Callable[[T], R]) -> R:
    return callback(world.get(dependency))


@injectable
class Customer:
    first_name: str = "John"


@injectable
@dataclass
class GreeterFirstName:
    customer_name: str = get(Customer, lambda c: c.first_name)
```

**Reply by pauleveritt** [reply]
In this example, is there any way in `get` (or just upstream or downstream) for it to know the parameter name of `customer_name`?

**Reply by Finistere** [reply]
You could with a descriptor and  `__set_name__()` but it won't be type-safe in the general case

```python
from antidote import injectable, world


class magic_get:
    __name: str

    def __init__(self, dependency: object) -> None:
        self.__dependency = dependency

    def __set_name__(self, owner, name):
        self.__name = name

    def __get__(self, instance, owner):
        assert instance is not None
        value = getattr(world[self.__dependency], self.__name)
        setattr(instance, self.__name, value)
        return value


@injectable
class Customer:
    first_name: str = "John"


@injectable
class GreeterFirstName:
    first_name: str = magic_get(Customer)


assert world[GreeterFirstName].first_name == world[Customer].first_name

```
